src/main/java/dan200/computercraft/shared/peripheral/speaker/SpeakerPeripheral.java

```diff
@@ -0,0 +1,250 @@
+        {
+            throw new LuaException("Invalid instrument, \"" + arguments[0] + "\"!");
+        }
+
+        if (arguments.length > 1)
+        {
+            if (!(arguments[1] instanceof Double))   // Arg wrong type
+            {
+                throw new LuaException("Expected string, number (optional), number (optional)");
+            }
+            volume = ((Double) arguments[1]).floatValue();
+
+        }
+
+        if (arguments.length > 2)
+        {
+            if (!(arguments[1] instanceof Double))  // Arg wrong type
+            {
+                throw new LuaException("Expected string, number (optional), number (optional)");
+            }
+            pitch = ((Double) arguments[2]).floatValue();
+        }
+
+        if (arguments.length > 3)
+        {
+            throw new LuaException("Expected string, number (optional), number (optional)");
+        }
+
+        // If the resource location for noteblock notes changes, this method call will need to be updated
+        Object[] returnValue = playSound(new Object[] {"block.note." + arguments[0], volume, Math.pow(2d, (pitch - 12) / 12d)}, context, true);
+
+        if (returnValue[0] instanceof Boolean && (Boolean) returnValue[0])
+        {
+            m_notesThisTick++;
+        }
+
+        return returnValue;
+
+    }
+
+    private Object[] playSound(Object[] arguments, ILuaContext context, boolean isNote) throws LuaException
+    {
+
+        float volume = 1f;
+        float pitch = 1f;
+
+        // Check if arguments are correct
+        if (arguments.length == 0) // Too few args
+        {
+            throw new LuaException("Expected string, number (optional), number (optional)");
+        }
+
+        if (!(arguments[0] instanceof String)) // Arg wrong type
+        {
+            throw new LuaException("Expected string, number (optional), number (optional)");
+        }
+
+        if (arguments.length > 1)
+        {
+            if (!(arguments[1] instanceof Double))  // Arg wrong type
+            {
+                throw new LuaException("Expected string, number (optional), number (optional)");
+            }
+            volume = ((Double) arguments[1]).floatValue();
+
+        }
+
+        if (arguments.length > 2)
+        {
+            if (!(arguments[2] instanceof Double))  // Arg wrong type
+            {
+                throw new LuaException("Expected string, number (optional), number (optional)");
+            }
+            pitch = ((Double) arguments[2]).floatValue();
+        }
+
+        if (arguments.length > 3)
+        {
+            throw new LuaException("Expected string, number (optional), number (optional)");
+        }
+
+        ResourceLocation resourceName = new ResourceLocation((String) arguments[0]);
+
+        if (m_clock - m_lastPlayTime >= TileSpeaker.MIN_TICKS_BETWEEN_SOUNDS || ((m_clock - m_lastPlayTime == 0) && (m_notesThisTick < ComputerCraft.maxNotesPerTick) && isNote))
```

Comment:
Yes. You can probably mark the entire method as `synchronized`.  I don't know if `TilePeirpheral.update` also needs to lock though.